File: javascript/packages/analysis/test/view-roots.test.ts
import { describe, test, expect } from "vitest"

import {
  partialNameForRoots,
  relativeToViewRoots,
  resolvePartial,
  rootIndexFor,
  templateNameForRoots,
} from "../src/partial-resolution"

import type { PartialPaths } from "../src/partial-resolution"

const APP = "app/views"
const ENGINE = "engines/billing/app/views"
const ROOTS = [APP, ENGINE]

describe("relativeToViewRoots", () => {
  test("returns the first root that contains the file", () => {
    expect(relativeToViewRoots(`${APP}/home/index.html.erb`, ROOTS)).toEqual([0, "home/index.html.erb"])
    expect(relativeToViewRoots(`${ENGINE}/billing/_invoice.html.erb`, ROOTS)).toEqual([1, "billing/_invoice.html.erb"])
  })

  test("returns null when no root contains the file", () => {
    expect(relativeToViewRoots("lib/elsewhere/_thing.html.erb", ROOTS)).toBeNull()
  })
})

describe("partialNameForRoots", () => {
  test("names a partial from a secondary view root", () => {
    expect(partialNameForRoots(`${ENGINE}/billing/_invoice.html.erb`, ROOTS)).toBe("billing/invoice")
  })

  test("names a partial from the primary view root", () => {
    expect(partialNameForRoots(`${APP}/shared/_header.html.erb`, ROOTS)).toBe("shared/header")
  })

  test("returns null for a file outside every root", () => {
    expect(partialNameForRoots("lib/_thing.html.erb", ROOTS)).toBeNull()
  })
})

describe("templateNameForRoots", () => {
  test("names a template from a secondary view root", () => {
    expect(templateNameForRoots(`${ENGINE}/billing/index.html.erb`, ROOTS)).toBe("billing/index")
  })
})

describe("rootIndexFor", () => {
  test("orders an earlier view root ahead of a later one", () => {
    expect(rootIndexFor(`${APP}/billing/_invoice.html.erb`, ROOTS)).toBe(0)
    expect(rootIndexFor(`${ENGINE}/billing/_invoice.html.erb`, ROOTS)).toBe(1)
  })

  test("sorts an unknown file last", () => {
    expect(rootIndexFor("lib/_thing.html.erb", ROOTS)).toBe(ROOTS.length)
  })
})

describe("resolvePartial", () => {
  test("resolves a sibling within the root that owns the caller", () => {
    const index: PartialPaths = new Map([["billing/row", `${ENGINE}/billing/_row.html.erb`]])
    const caller = `${ENGINE}/billing/index.html.erb`

    expect(resolvePartial("row", caller, index, ROOTS)).toBe(`${ENGINE}/billing/_row.html.erb`)
  })

  test("resolves with a single root", () => {
    const index: PartialPaths = new Map([["shared/header", `${APP}/shared/_header.html.erb`]])

    expect(resolvePartial("shared/header", "", index, [APP])).toBe(`${APP}/shared/_header.html.erb`)
  })
})

File: test/analysis/view_roots_test.rb
# frozen_string_literal: true

require_relative "../test_helper"
require_relative "../../lib/herb/analysis/partial_index"

require "tmpdir"

module Analysis
  class ViewRootsTest < Minitest::Spec
    def write(root, relative, body = "<div></div>\n")
      path = File.join(root, relative)

      FileUtils.mkdir_p(File.dirname(path))
      File.write(path, body)

      path
    end

    test "names a partial from a secondary view root" do
      Dir.mktmpdir do |dir|
        app = File.join(dir, "app", "views")
        engine = File.join(dir, "engines", "billing", "app", "views")

        entry = write(app, "home/index.html.erb")
        invoice = write(engine, "billing/_invoice.html.erb")

        index = Herb::Analysis::PartialIndex.new([app, engine], [entry, invoice])

        assert_equal ["billing/invoice"], index.names
      end
    end

    test "an earlier view root shadows a later one" do
      Dir.mktmpdir do |dir|
        app = File.join(dir, "app", "views")
        engine = File.join(dir, "engines", "billing", "app", "views")

        engine_invoice = write(engine, "billing/_invoice.html.erb", "<div>engine</div>\n")
        app_invoice = write(app, "billing/_invoice.html.erb", "<div>app</div>\n")

        index = Herb::Analysis::PartialIndex.new([app, engine], [engine_invoice, app_invoice])
        resolved = index.resolve("billing/invoice", nil)

        assert_equal 2, resolved.size
        assert_equal app_invoice, resolved.first
      end
    end

    test "resolves a sibling within the root that owns the caller" do
      Dir.mktmpdir do |dir|
        app = File.join(dir, "app", "views")
        engine = File.join(dir, "engines", "billing", "app", "views")

        entry = write(engine, "billing/index.html.erb")
        row = write(engine, "billing/_row.html.erb")

        index = Herb::Analysis::PartialIndex.new([app, engine], [entry, row])

        assert_equal [row], index.resolve("row", entry)
      end
    end

    test "a single root still resolves" do
      Dir.mktmpdir do |dir|
        app = File.join(dir, "app", "views")
        header = write(app, "shared/_header.html.erb")

        index = Herb::Analysis::PartialIndex.new([app], [header])

        assert_equal ["shared/header"], index.names
        assert_equal [header], index.resolve("shared/header", nil)
      end
    end
  end
end
